File: src/main/java/com/plugin/frege/psi/impl/FregePsiMethodImpl.java
package com.plugin.frege.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.*;
import com.intellij.psi.impl.light.LightParameter;
import com.intellij.psi.impl.light.LightParameterListBuilder;
import com.intellij.psi.impl.light.LightReferenceListBuilder;
import com.intellij.psi.impl.light.LightTypeElement;
import com.intellij.psi.javadoc.PsiDocComment;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.util.MethodSignature;
import com.intellij.psi.util.MethodSignatureBackedByPsiMethod;
import com.intellij.psi.util.PsiTreeUtil;
import com.plugin.frege.FregeLanguage;
import com.plugin.frege.psi.FregeBinding;
import com.plugin.frege.psi.FregePsiMethod;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

@SuppressWarnings("UnstableApiUsage")
public abstract class FregePsiMethodImpl extends FregeNamedElementImpl implements FregePsiMethod {
    static private PsiType objectType = null;
    private LightTypeElement objectTypeElement = null;

    public FregePsiMethodImpl(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public @NotNull PsiReferenceList getThrowsList() {
        return new LightReferenceListBuilder(getManager(), FregeLanguage.INSTANCE, PsiReferenceList.Role.THROWS_LIST);
    }

    @Override
    public boolean isVarArgs() {
        return false;
    }

    @Override
    public PsiMethod @NotNull [] findSuperMethods() {
        return PsiMethod.EMPTY_ARRAY; // TODO
    }

    @Override
    public PsiMethod @NotNull [] findSuperMethods(boolean checkAccess) {
        return PsiMethod.EMPTY_ARRAY; // TODO
    }

    @Override
    public PsiMethod @NotNull [] findSuperMethods(PsiClass parentClass) {
        return PsiMethod.EMPTY_ARRAY; // TODO
    }

    @Override
    public @NotNull List<MethodSignatureBackedByPsiMethod> findSuperMethodSignaturesIncludingStatic(boolean checkAccess) {
        return List.of(); // TODO
    }

    @Override
    public @Nullable PsiMethod findDeepestSuperMethod() {
        return null; // TODO
    }

    @Override
    public PsiMethod @NotNull [] findDeepestSuperMethods() {
        return PsiMethod.EMPTY_ARRAY; // TODO
    }

    @Override
    public boolean isDeprecated() {
        return false; // TODO
    }

    @Override
    public @Nullable PsiDocComment getDocComment() {
        return null; // TODO
    }

    @Override
    public @Nullable PsiClass getContainingClass() {
        return FregePsiClassUtilImpl.findContainingFregeClass(this);
    }

    @Override
    public boolean hasTypeParameters() {
        return false; // Unless we want to support generics
    }

    @Override
    public @Nullable PsiTypeParameterList getTypeParameterList() {
        return null; // Unless we want to support generics
    }

    @Override
    public PsiTypeParameter @NotNull [] getTypeParameters() {
        return PsiTypeParameter.EMPTY_ARRAY; // Unless we want to support generics
    }

    @Override
    public @NotNull PsiParameterList getParameterList() {
        LightParameterListBuilder list = new LightParameterListBuilder(getManager(), FregeLanguage.INSTANCE);
        int paramsNumber = getParamsNumber();
        PsiType object = Objects.requireNonNull(getObjectType());
        PsiElement scope = Objects.requireNonNull(PsiTreeUtil.getParentOfType(this, FregeBinding.class));
        for (int i = 0; i < paramsNumber; i++) {
            list.addParameter(new LightParameter("arg" + i, object, scope));
        }

        return list; // TODO a normal type system
    }

    @Override
    public @NotNull MethodSignature getSignature(@NotNull PsiSubstitutor substitutor) {
        return MethodSignatureBackedByPsiMethod.create(this, substitutor);
    }

    @Override
    public boolean hasModifierProperty(@NonNls @NotNull String name) {
        return getModifierList().hasModifierProperty(name);
    }

    @Override
    public @NotNull String getName() {
        return Objects.requireNonNull(super.getName()); // only for making it @NotNull
    }

    protected abstract int getParamsNumber();

    protected @Nullable  PsiType getObjectType() {
        if (objectType == null) {
            objectType = PsiMethodReferenceType.getJavaLangObject(getManager(),
                    GlobalSearchScope.everythingScope(getProject()));
        }
        return objectType;
    }

    protected @Nullable PsiTypeElement getObjectTypeElement() {
        if (objectTypeElement == null) {
            PsiType object = getObjectType();
            if (object != null) {
                objectTypeElement = new LightTypeElement(getManager(), object);
            }
        }
        return objectTypeElement;
    }
}

File: src/main/kotlin/com/plugin/frege/psi/impl/FregePsiMethodImpl.kt
package com.plugin.frege.psi.impl

import com.intellij.lang.ASTNode
import com.intellij.psi.*
import com.intellij.psi.impl.light.*
import com.intellij.psi.impl.source.HierarchicalMethodSignatureImpl
import com.intellij.psi.javadoc.PsiDocComment
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.util.MethodSignature
import com.intellij.psi.util.MethodSignatureBackedByPsiMethod
import com.plugin.frege.FregeLanguage
import com.plugin.frege.psi.FregePsiClass
import com.plugin.frege.psi.FregePsiMethod
import com.plugin.frege.resolve.FregeResolveUtil
import com.plugin.frege.stubs.FregeMethodStub
import org.jetbrains.annotations.NonNls

@Suppress("UnstableApiUsage")
abstract class FregePsiMethodImpl : FregeNamedStubBasedPsiElementBase<FregeMethodStub>, FregePsiMethod {
    constructor(node: ASTNode) : super(node)

    constructor(stub: FregeMethodStub, nodeType: IStubElementType<*, *>) : super(stub, nodeType)

    private val objectType: PsiType by lazy {
        PsiMethodReferenceType.getJavaLangObject(manager, GlobalSearchScope.everythingScope(project))
    }

    protected val objectTypeElement: LightTypeElement? by lazy {
        LightTypeElement(manager, objectType)
    }

    override fun onlyQualifiedSearch(): Boolean = false

    override fun getThrowsList(): PsiReferenceList =
        LightReferenceListBuilder(manager, FregeLanguage.INSTANCE, PsiReferenceList.Role.THROWS_LIST)

    override fun isVarArgs(): Boolean = false

    override fun findSuperMethods(): Array<PsiMethod> = PsiMethod.EMPTY_ARRAY // TODO

    override fun findSuperMethods(checkAccess: Boolean): Array<PsiMethod> = PsiMethod.EMPTY_ARRAY // TODO

    override fun findSuperMethods(parentClass: PsiClass): Array<PsiMethod> = PsiMethod.EMPTY_ARRAY // TODO

    override fun findSuperMethodSignaturesIncludingStatic(
        checkAccess: Boolean
    ): List<MethodSignatureBackedByPsiMethod> = emptyList() // TODO

    override fun findDeepestSuperMethod(): PsiMethod? = null // TODO

    override fun findDeepestSuperMethods(): Array<PsiMethod> = PsiMethod.EMPTY_ARRAY // TODO

    override fun isDeprecated(): Boolean = false // TODO

    override fun getDocComment(): PsiDocComment? = null // TODO

    override fun getContainingClass(): FregePsiClass? = FregeResolveUtil.findContainingFregeClass(this)

    override fun hasTypeParameters(): Boolean = false // Unless we want to support generics

    override fun getTypeParameterList(): PsiTypeParameterList? = null // Unless we want to support generics

    override fun getTypeParameters(): Array<PsiTypeParameter> =
        PsiTypeParameter.EMPTY_ARRAY // Unless we want to support generics

    protected abstract fun getParamsNumber(): Int

    override fun getParameterList(): PsiParameterList {
        val list = LightParameterListBuilder(manager, FregeLanguage.INSTANCE)
        val paramsNumber = getParamsNumber()
        for (i in 0 until paramsNumber) {
            list.addParameter(LightParameter("arg$i", objectType, this))
        }
        return list // TODO type inference
    }

    override fun getSignature(substitutor: PsiSubstitutor): MethodSignature =
        MethodSignatureBackedByPsiMethod.create(this, substitutor)

    override fun getModifierList(): LightModifierList {
        val baseList = LightModifierList(manager, FregeLanguage.INSTANCE, PsiModifier.FINAL)
        baseList.addModifier(accessPsiModifier)
        return baseList
    }

    override fun hasModifierProperty(@NonNls name: String): Boolean = modifierList.hasModifierProperty(name)

    override fun getName(): String = greenStub?.name ?: nameIdentifier?.text ?: text

    override fun getHierarchicalMethodSignature(): HierarchicalMethodSignature {
        return HierarchicalMethodSignatureImpl(
            MethodSignatureBackedByPsiMethod.create(this, PsiSubstitutor.EMPTY)
        )
    }

    override fun getReturnType(): PsiType = objectType // TODO waiting for type system

    override fun getReturnTypeElement(): PsiTypeElement? = objectTypeElement // TODO waiting for type system
}
